panther_analysis_tool/main.py

```diff
@@ -1122,115 +1149,96 @@ def _run_tests(
-                f"{str(VALID_SEVERITIES)}, got [{str(function_output)}] instead."
-            )
-    elif function_name == "alert_context":
-        is_valid = isinstance(function_output, dict)
-    elif function_name == "destinations":
-        # Type checking for destinations
-        if isinstance(function_output, list):
-            # Type checking for elements in list
-            # Note: we cannot perform value validation unless we were to query the outputs-api
-            if not all(isinstance(x, str) for x in function_output):
-                # List contains non-string element, so enumerate the types present for error msg
-                list_types = set()
-                for each_item in function_output:
-                    list_types.add(type(each_item))
-                is_valid = False
-                output = AttributeError(
-                    f"Expected [{function_name}] to return a list of strings, "
-                    f"got types [{list_types}] instead"
-                )
-        else:
-            is_valid = False
-    # For reserved functions besides severity and destinations, we expect a string output
+    """Determine whether an auxiliary function for a rule raised an error"""
+    function_error = getattr(rule_result, f"{function_name}_exception")
+    output = getattr(rule_result, f"{function_name}_output")
+    is_defined = getattr(rule_result, f"{function_name}_defined")
+    failed = function_error is not None
+
+    # For backwards compatibility we can accept invalid destination names,
+    # as long as a string is returned. Strict check is enabled when users
+    # pass destination names explicitly through command-line parameters.
+    if function_name == "destinations" and failed and not strict_check:
+        # Otherwise we fall back to a best-effort check of the return type only
+        if isinstance(function_error, UnknownDestinationError):
+            exc: UnknownDestinationError = function_error
+            failed = not _check_destinations_type(exc.result())
+
+    # The function has not been executed
+    if not is_defined:
+        display = None
     else:
-        is_valid = isinstance(function_output, str)
+        display = function_error if failed else output
+
+    return {
+        "error": function_error,
+        "output": output,
+        "failed": failed,
+        "has_invalid_type": isinstance(function_error, FunctionReturnTypeError),
+        "display": display,
+    }
 
-    # Invalid Types, used to generate a corresponding error message
-    if not is_valid and output == function_output:
-        if function_name == "destinations":
-            expected_type = "list"
-        else:
-            expected_type = "string"
 
-        output = AttributeError(
-            f"Expected [{function_name}] to return a [{expected_type}], "
-            f"got [{type(function_output)}] instead"
-        )
-    return is_valid, output
+def _check_destinations_type(obj: Any) -> bool:
```

Comment:
This check should normally be provided by the `rule` module (or the `Rule` class). I didn't want to make further changes to the `Rule` class at this point to avoid making the PR more complex.